massa-execution-worker/src/speculative_roll_state.rs

```diff
@@ -148,6 +149,61 @@ impl SpeculativeRollState {
         Ok(())
     }
 
+    /// Try to slash `roll_count` rolls from the given address. If not enough roll, slash
+    /// the available amount and return the value.
+    ///
+    /// # Arguments
+    /// * `addr`: address to slash the rolls from
+    /// * `roll_count`: number of rolls to slash
+    pub fn try_slash_rolls(
+        &mut self,
+        addr: &Address,
+        roll_count: u64,
+    ) -> Result<u64, ExecutionError> {
+        // fetch the roll count
+        let owned_count = self.get_rolls(addr);
+        let roll_to_slash = min(roll_count, owned_count);
+
+        match self.added_changes.roll_changes.get_mut(addr) {
+            None => {
+                // Address does not exist anymore, slash nothing
+                Ok(0)
+            }
+            Some(current_rolls) => {
+                *current_rolls = current_rolls.saturating_sub(roll_to_slash);
+                Ok(roll_to_slash)
+            }
+        }
+    }
+
+    /// Try to slash `amount` credits from the given address. If not enough credits, slash
+    /// the available amount and return the value.
+    ///
+    /// # Arguments
+    /// * `addr`: address to slash the deferred credits from
+    /// * `amount`: number of deferred credits to slash
+    pub fn try_slash_deferred_credits(
+        &mut self,
+        slot: &Slot,
+        addr: &Address,
+        amount: &Amount,
+    ) -> Amount {
+        let credits = self.get_address_deferred_credits(addr, *slot);
+        let mut remaining_to_slash = *amount;
+
+        for (slot, amount) in credits.iter() {
+            let to_slash = min(*amount, remaining_to_slash);
+            let new_deferred_credits = amount.saturating_sub(to_slash);
+            remaining_to_slash = remaining_to_slash.saturating_sub(to_slash);
+
+            self.added_changes
+                .deferred_credits
+                .insert(*slot, *addr, new_deferred_credits);
+        }
+
+        remaining_to_slash
```

Comment:
to make it uniform with try_slash_rolls, maybe let's make both functions return either the amount slashed or the remaining amount

Reply:
Fixed. See below